src/os/layer/preset/presetmenubutton.js

```diff
@@ -0,0 +1,220 @@
+
+const LayerPresetManager = goog.require('os.layer.preset.LayerPresetManager');
+const MenuButtonCtrl = goog.require('os.ui.menu.MenuButtonCtrl');
+const Menu = goog.require('os.ui.menu.Menu');
+const MenuItem = goog.require('os.ui.menu.MenuItem');
+const MenuItemType = goog.require('os.ui.menu.MenuItemType');
+const Module = goog.require('os.ui.Module');
+const {Presets: Metrics} = goog.require('os.metrics.keys');
+
+
+/**
+ * @const {string}
+ */
+const MENU_FLAG = 'presets';
+
+/**
+ * Preset events
+ * @enum {string}
+ */
+const EventType = {
+  REMOVE: 'remove',
+  SAVE: 'save',
+  TOGGLE_DEFAULT_TRUE: 'toggle-default-true',
+  TOGGLE_DEFAULT_FALSE: 'toggle-default-false',
+  TOGGLE_PUBLISHED_TRUE: 'toggle-published-true',
+  TOGGLE_PUBLISHED_FALSE: 'toggle-published-false'
+};
+
+/**
+ * The controller for the preset directive; make use of the MenuButtonController
+ * @unrestricted
+ */
+class Controller extends MenuButtonCtrl {
+  /**
+   * Constructor.
+   * @param {!angular.Scope} $scope
+   * @param {!angular.JQLite} $element The element
+   * @ngInject
+   */
+  constructor($scope, $element) {
+    super($scope, $element);
+
+    /**
+     * @type {boolean}
+     */
+    this['isAdmin'] = false;
+
+    this.init_();
+  }
+
+  /**
+   * @private
+   */
+  init_() {
+    this['isAdmin'] = LayerPresetManager.getInstance().isAdmin(); // requires ng-if
+
+    // build the menu
+    this.menu = new Menu(new MenuItem({
+      type: MenuItemType.ROOT,
+      children: [{
+        label: '',
+        type: MenuItemType.GROUP,
+        sort: 0,
+        beforeRender: this.visibility.bind(this),
+        children: [{
+          label: 'Save as...',
+          eventType: EventType.SAVE,
+          tooltip: 'Save Preset',
+          icons: ['<i class="fa fa-fw fa-floppy-o"></i>'],
+          handler: this.save.bind(this),
+          metricKey: Metrics.SAVE,
+          sort: 0
+        }, {
+          label: 'Publish',
+          eventType: EventType.TOGGLE_PUBLISHED_TRUE,
+          tooltip: 'Make Preset visible to everyone',
+          icons: ['<i class="fa fa-fw fa-eye"></i>'],
+          handler: this.togglePublished.bind(this),
+          metricKey: Metrics.TOGGLE_PUBLISHED,
+          sort: 10
+        }, {
+          label: 'Draft',
+          eventType: EventType.TOGGLE_PUBLISHED_FALSE,
+          tooltip: 'Return Preset to draft mode',
+          icons: ['<i class="fa fa-fw fa-eye-slash"></i>'],
+          handler: this.togglePublished.bind(this),
+          metricKey: Metrics.TOGGLE_PUBLISHED,
+          sort: 11
+        }, {
+          label: 'Set as Default',
+          eventType: EventType.TOGGLE_DEFAULT_TRUE,
+          tooltip: 'Make Preset the default style when layer added',
+          icons: ['<i class="fa fa-fw fa-star"></i>'],
+          handler: this.toggleDefault.bind(this),
+          metricKey: Metrics.TOGGLE_DEFAULT,
+          sort: 20
+        }, {
+          label: 'Revoke Default',
```

Comment:
Revoke and Set as Default being separate menu options seems confusing. It doesn't look like they have a visibility toggle, so both will show at all times regardless of whether the selected option is already the default. It would be better for this to be a single option that changes its text and functionality based on the state of the underlying preset. The `beforeRender` function can change things like the `label` property as well I believe.

Reply:
There's a beforeRender on the group which shows/hides the MenuItems as appropriate for the selected Preset -- it does so in one handler rather than multiple handlers. I can edit that handler to swap around the icons, labels, and events if we have to have less MenuItems; but this is simple and effective.